### ISSUE TITLE
Registering a global fallback for all operators that defaults us to assuming that autograd is not implemented

### ISSUE THREAD

**Issue by soulitzer** [pytorch/pytorch#65151]
## 🚀 Feature

As originally pointed out here: https://github.com/pytorch/pytorch/issues/62032

> The boxed fallback has an added benefit that we could then consider, at a later point in time, to register it as the universal Autograd fallback key, thus automatically marking all custom operators as not supporting autograd unless they explicitly register an Autograd implementation (this would be quite BC breaking and there would be quite a bit of fallout to clean up afterwards).

## Pitch

The first part of this task is to investigate the fallout that we would need to clean up

Some notes from discussion with @ezyang and @albanD
1. Many third-party operators may (incorrectly) rely on the operator to be composite even though it is registered to be non-composite (e.g. CPU/CUDA)
   - In these cases, we should ask user should actually register as composite
 2. There will be a perf hit for certain cases (the user would have to realize this themselves):
    - Using inference mode may help bypass the new Autograd kernel (but may also hurt in some cases because we would no longer skip the ADInplaceOrView kernel, which does some FunctionSchema checking).
    - In the cases where inference mode may hurt, the original workaround in the manual case is that the user should just not register the ADInplaceOrView kernel. How would that translate here?

cc @ezyang @gchanan @albanD @gqchen @nikitaved @Varal7 @xmfan @chauhang @penguinwu @bdhirsh @zou3519 @pearu @soulitzer @Lezcano

**Comment by ezyang** [comment]
Here is a refreshed description for 2025:

A long-standing mistake we made with the custom ops API which is difficult to fix now is that if you define an operator with CPU/CUDA key only, we will silently drop gradients if you try to differentiate through it (as there is no autograd registration). The history here is that Jeffrey Wan originally added the autograd fallback implementation in 2021 (https://github.com/pytorch/pytorch/pull/63458) and Richard Zou formally deprecated it in 2023 (https://github.com/pytorch/pytorch/pull/104481). At both points in time, we were scared of silently breaking people. However, the year is now 2025 and I think we should consider actually pulling the plug.

I think a reasonable start is to start enforcing Autograd registrations as seen in https://github.com/pytorch/pytorch/pull/166367 . Actually, I had not realized the change I proposed in this diff was technically BC breaking; I thought this PR was necessary to detect a silent correctness error divergence between eager and autograd (in fact, we ALREADY had detected the problem that precipitated this PR, but the opcheck test was xfailed for an unrelated reason.) But having gone through a few revisions of this diff, I have found it is not so BC breaking in fbcode, and so I think we should seriously consider landing it (with some care!) I have found only one op in internal, fb::conditional, which is for some reason being backpropped through without an explicit Autograd registration, and I suspect this is a bug (@mengluy  to follow up here)

Beyond this, I think we should also actually start enforcing this in eager. I think OSS will be able to deal. Here is a repro for us NOT erroring today:

```
import torch
from torch.utils.cpp_extension import load_inline

cpp_source = """
#include <torch/extension.h>

torch::Tensor copy_op_cpu(const torch::Tensor& input) {
  auto output = torch::empty_like(input);
  std::memcpy(output.data_ptr(), input.data_ptr(), input.nbytes());
  return output;
}

TORCH_LIBRARY(mylib, m) {
  m.def("copy_op(Tensor input) -> Tensor");
}

TORCH_LIBRARY_IMPL(mylib, CPU, m) {
  m.impl("copy_op", copy_op_cpu);
}
"""

load_inline(
    name="mylib",
    cpp_sources=cpp_source,
    functions=["copy_op_cpu"],
    verbose=True,
)

x = torch.randn(3, 4, requires_grad=True)
y = torch.ops.mylib.copy_op(x)
y.sum().backward()

```

**Comment by albanD** [comment]
The next step we could do for sure is make these an error by default, with an env-variable + global flag to make it a warning again.
Which after a few years of warning feel fair.

The only concern is that, while this is bad, users can get the "correct" behavior in this case when the kernel at the CPU key actually has an autograd.Function. And this will only break if using advanced features like PT2, subclasses, etc.
But I think I agree that, in 2025, all custom ops should either work with PT2 autograd or run in inference_mode.
